Tools/clinic/clinic.py

```diff
@@ -922,41 +922,37 @@ def deprecate_positional_use(
             params: dict[int, Parameter],
     ) -> str:
         assert len(params) > 0
-        names = [repr(p.name) for p in params.values()]
-        first_pos, first_param = next(iter(params.items()))
-        last_pos, last_param = next(reversed(params.items()))
-
-        # Pretty-print list of names.
-        pstr = pprint_words(names)
-
-        # For now, assume there's only one deprecation level.
-        assert first_param.deprecated_positional == last_param.deprecated_positional
-        thenceforth = first_param.deprecated_positional
-        assert thenceforth is not None
-        major, minor = thenceforth
+        first_pos = next(iter(params))
+        last_pos = next(reversed(params))
 
         # Format the deprecation message.
-        if first_pos == 0:
-            preamble = "Passing positional arguments to "
         if len(params) == 1:
             condition = f"nargs == {first_pos+1}"
-            if first_pos:
-                preamble = f"Passing {first_pos+1} positional arguments to "
-            message = preamble + (
-                f"{func.fulldisplayname}() is deprecated. Parameter {pstr} will "
-                f"become a keyword-only parameter in Python {major}.{minor}."
-            )
+            amount = f"{first_pos+1} " if first_pos else ""
+            pl = "s"
         else:
             condition = f"nargs > {first_pos} && nargs <= {last_pos+1}"
-            if first_pos:
-                preamble = (
-                    f"Passing more than {first_pos} positional "
-                    f"argument{'s' if first_pos != 1 else ''} to "
+            amount = f"more than {first_pos} " if first_pos else ""
+            pl = "s" if first_pos != 1 else ""
+        message = (
+            f"Passing {amount}positional argument{pl} to "
+            f"{func.fulldisplayname}() is deprecated."
+        )
+
+        for (major, minor), group in itertools.groupby(params.values(),  # type: ignore[misc]
+                                        lambda p: p.deprecated_positional):
```

Comment:
I'd find these uses of `groupby` more readable if we did this:

```diff
--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -35,6 +35,7 @@
     Iterator,
     Sequence,
 )
+from operator import attrgetter
 from types import FunctionType, NoneType
 from typing import (
     TYPE_CHECKING,
@@ -939,8 +940,9 @@ def deprecate_positional_use(
             f"{func.fulldisplayname}() is deprecated."
         )

-        for (major, minor), group in itertools.groupby(params.values(),  # type: ignore[misc]
-                                        lambda p: p.deprecated_positional):
+        for (major, minor), group in itertools.groupby(
+            params.values(), key=attrgetter("deprecated_positional")
+        ):
             names = [repr(p.name) for p in group]
             pstr = pprint_words(names)
             if len(names) == 1:
@@ -1006,8 +1008,9 @@ def deprecate_keyword_use(
             f"{func.fulldisplayname}() is deprecated."
         )

-        for (major, minor), group in itertools.groupby(params.values(),  # type: ignore[misc]
-                                        lambda p: p.deprecated_keyword):
+        for (major, minor), group in itertools.groupby(
+            params.values(), key=attrgetter("deprecated_positional")
+        ):
             names = [repr(p.name) for p in group]
             pstr = pprint_words(names)
             pl = 's' if len(names) != 1 else ''
```

It also means we can get rid of a `type: ignore` (though just because mypy doesn't have a very good understanding of `attrgetter`, I think ;)

Reply:
I considered this idea, but abandoned it because it did not make the code shorter.